### PR TITLE
tried to figure out how to add more tests (this time checking if the …

### PR DESCRIPTION
**Pull request by marekjez86**
…whole range of a variable font is covered by XML -- we should)

added couple more fonts to XML

### CODE REVIEW

File: tests/noto_fonts_for_android_test.py [hunk 1]
```diff
@@ -121,5 +121,31 @@ def test_font_weights():
...
+      reached_min_weight.add(_font_file(font_el))
+    if xml_weight == max_weight:
+      reached_max_weight.add(_font_file(font_el))
+
+  print("======> minimum not reached")
+  for font_file_name in all_fonts.difference(reached_min_weight):
+    print(font_file_name)
+  print("======> maximum not reached")
+  for font_file_name in all_fonts.difference(reached_max_weight):
+    print(font_file_name)
```
Comment (rsheeter) [inline tests/noto_fonts_for_android_test.py:147]: As written the test doesn't actually fail if you violate it's conditions.

Are you sure you wnat to reach the min/max? In looking at fonts.xml I suspect we want to hit all whole multiple of 100 between min and max weight?

I think that might lead me to write the test as something along the lines of:

```
# at the top
import collections

# in the test
for family in root.iter("family"):
  font_to_xml_weights = collections.defaultdict(set)
  for font in family.xpath("//font[@path]"):  # ...I think you can xpath within a 
    font_to_xml_weights[_font_path(font)).add(int(font_el.attrib["weight"]))

  # now you have a map of font path => set of weights in xml
  for font_path, xml_weights in font_to_xml_weights.items():
    # open the font, compute the 100 weights between it's min/max weight
   # if xml_weights != computed weights add this to the error list

  assert there are no errors, ", ".join(errors)
```

File: android-connection/noto-fonts-4-android.xml [hunk 2]
```diff
@@ -1804,4 +1804,27 @@
...
+		<font weight="600" style="normal" fallbackFor="serif" postScriptName="NotoSansTangsa-Regular" path="unhinted/slim-variable-ttf">
+	          NotoSansTangsa-VF.ttf
+	          <axis tag="wght" stylevalue="600"/>
+	      </font>
+		<font weight="700" style="normal" fallbackFor="serif" postScriptName="NotoSansTangsa-Regular" path="unhinted/slim-variable-ttf">
+	          NotoSansTangsa-VF.ttf
+	          <axis tag="wght" stylevalue="700"/>
+	      </font>
+	</family>
 </familyset>
```
Comment (rsheeter) [inline android-connection/noto-fonts-4-android.xml:1830]: since this is unrelated - I think? - to the noto_fonts_for_android_test.py change my suggestion would be to do it in separate PRs. That is, make another branch that's a copy of this one and revert changes to one of the files on each branch.

Comment (marekjez86) [reply android-connection/noto-fonts-4-android.xml:1830]: will do (re branch)

File: tests/noto_fonts_for_android_test.py (hunk 1 above)
Comment (marekjez86) [reply tests/noto_fonts_for_android_test.py:147]: Your code looks better. In my code, I wanted to check if we ever have this situation (just in case this omission was intentional; not exposing all the weights is not wrong if done purposefully [I have no idea for what purpose though :-)]). BTW, SansOriya and SansOriyaUI are in this situation and I will expand the XML to cover the whole range in these VFs.

Comment (rsheeter) [reply tests/noto_fonts_for_android_test.py:147]: You could have a list of expected exceptions to the rule perhaps? - that way it'll flag if, as I believe is normal, it's a mistake and we still have an opt-out.

I'm curious ... when would we want to NOT list all the weights?

**Comment by marekjez86** [comment]
I didn't know how to revert in my command line so I just did it via a brute force

File: tests/noto_fonts_for_android_test.py [hunk 3]
```diff
@@ -121,5 +132,25 @@ def test_font_weights():
   assert not errors, ", ".join(errors)
 
 
+def test_font_full_weight_coverage():
+  root = etree.parse(str(_noto_4_android_file()))
+  errors = []
+  for family in root.iter("family"):
+    font_to_xml_weights = collections.defaultdict(set)
+    for font in family.xpath("//font[@path]"): 
+      font_to_xml_weights[_font_path(font)].add(int(font.attrib["weight"]))
```
Comment (rsheeter) [inline tests/noto_fonts_for_android_test.py:141]: you can stuff the index into the key here so it's a tuple of (path, index)`font_to_xml_weights[(_font_path(font), font.attrib.get("index", -1))].add(int(font.attrib["weight"]))`

File: tests/noto_fonts_for_android_test.py [hunk 4]
```diff
@@ -121,5 +132,25 @@ def test_font_weights():
...
+
+  # now you have a map of font path => set of weights in xml
+  for font_path, xml_weights in font_to_xml_weights.items():
```
Comment (rsheeter) [inline tests/noto_fonts_for_android_test.py:144]: once you have index in the map key you can capture it here to pass in to _open_font_path

`for (font_path, font_number), xml_weights in font_to_xml_weights.items():`

File: tests/noto_fonts_for_android_test.py [hunk 5]
```diff
@@ -52,6 +53,16 @@ def _open_font(font_el) -> ttLib.TTFont:
   return ttLib.TTFont(str(path))
 
 
+def _open_font_path(path) -> ttLib.TTFont:
+  if not path.is_file():
+    raise IOError(f"No such file: {path}")
+
+  if str(path).lower().endswith(".ttc"):
+    # I had no idea how to get fontNumber out of a path :-) so I fixed it at 1
+    return ttLib.TTFont(str(path), fontNumber=1)
```
Comment (rsheeter) [inline tests/noto_fonts_for_android_test.py:62]: make font_number an argument IMO

**Comment by rsheeter** [comment]
> I didn't know how to revert XML in my command line

you want to do something like `git rebase -i HEAD~6` then edit the "pick" beside the commit syou don't want to say "drop" then force push to this branch. The unwanted commits will disappear. It's probably wise to make a branch that is a copy of this branch before you start, just in case you drop something you wanted to keep :)

I would make just the test changesone PR and the publishing new fonts and modifying the XML another PR (or PRs if the tasks are unrelated)

File: tests/noto_fonts_for_android_test.py (hunk 5 above)
Comment (rsheeter) [reply tests/noto_fonts_for_android_test.py]: Or you could do as I did here and just pass the font element in, https://github.com/googlefonts/noto-fonts/blob/main/tests/noto_fonts_for_android_test.py#L45. Actually ... isnt' that the same file? You could probably use that function?

File: tests/noto_fonts_for_android_test.py [hunk 6]
```diff
@@ -121,5 +130,25 @@ def test_font_weights():
...
+  root = etree.parse(str(_noto_4_android_file()))
+  errors = []
+  for family in root.iter("family"):
+    font_to_xml_weights = collections.defaultdict(set)
+    for font in family.xpath("//font[@path]"): 
+      font_to_xml_weights[(_font_path(font), font.attrib.get("index", -1))].add(int(font.attrib["weight"]))
+
+  # now you have a map of font path => set of weights in xml
+  for (font_path, font_number), xml_weights in font_to_xml_weights.items():
+  # open the font, compute the 100 weights between it's min/max weight
```
Comment (rsheeter) [inline tests/noto_fonts_for_android_test.py:143]: the 100-weights bit isn't happening, should revise the comment?

File: tests/noto_fonts_for_android_test.py [hunk 7]
```diff
@@ -121,5 +130,25 @@ def test_font_weights():
...
+  # if xml_weights != computed weights add this to the error list
+    font = _open_font_path(font_path, font_number)
+    min_wght, default_wght, max_weight = _weight(font)
+    if min(xml_weights) > min_wght or max(xml_weights) < max_weight:
+      errors.append(f"{font_path} weight range {min(xml_weights)}..{max(xml_weights)} could be expanded to {min_wght}..{max_weight}")
```
Comment (rsheeter) [inline tests/noto_fonts_for_android_test.py:148]: should or even must instead of "could"?

**Review by rsheeter — APPROVED** [review]
Approved though I still recommend considering moving the font file updates to their own PR

EDIT: also the tests report failing so we should make sure they pass before merging

**Comment by marekjez86** [comment]
They (font updates) will be separate. My mistake. Thank you.
